### PR TITLE
CmpName: Refactored wildcard processing to use backtracking instead of recursion.

### PR DESCRIPTION
**Pull request by yegor-mialyk**
## Summary

Refactored wildcard processing to use backtracking instead of recursion. Initially, I was motivated by #988 and the `BUGBUG rewrite` comment in the code.

 
## Checklist
- [x] I have followed the [contributing guidelines](https://github.com/FarGroup/FarManager/blob/master/CONTRIBUTING.md).
- [ ] I have discussed this with project maintainers:  <br/>
If not checked, I accept that this work might be rejected in favor of a different grand plan.<br/>

## Details

My first version of the CmpName algorithm was much stricter and more precise to my liking, but after finding the log below, I relaxed the logic and tried to replicate the current behavior as much as possible.

It passes all existing unit tests, and I have added many more. However, in some edge cases or with crazy patterns, it may produce slightly different results.

I manually tested hundreds of other patterns, comparing `cmd`, the current Far algorithm, and the new one and found that trying to mimic `cmd` is not a good idea. How in the world can `cmd`  match `t??????` to `test`? :)

Especially when Microsoft is promoting PowerShell, which is much stricter and more accurate than both `cmd` and the current Far algorithm (and a new one as it is also "relaxed").

```
drkns 2018-09-30 00:16:35+03:00 - build 5274

[...]

2. Correction of wildcards processing in file names. It should be closer to cmd now.
```

P. S. Any feedback is welcome. It is fine if we decide not to merge due to backward compatibility or other concerns.

Hope the code is easy to read and understand. But if it were my decision, I would move toward PowerShell compatibility rather than toward `cmd`.

### CODE REVIEW

File: far/processname.cpp [hunk 1]
```diff
@@ -144,142 +144,132 @@ string ConvertWildcards(string_view const SrcName, string_view const Mask)
...
-				if (pattern.size() == 2 && pattern[1]==L'*')
-					return true;
+		return false;
+	};
 
-				if (std::ranges::none_of(pattern, [](wchar_t Char) { return contains(L"*?["sv, Char); }))
-				{
-					const auto RDotIt = std::ranges::find(str | std::views::reverse, L'.');
-					const auto DotIt = RDotIt == str.crend()? str.cend() : (RDotIt + 1).base();
+	while (si <= str.size())
```
[the remark this answers is not in the document]
Comment (yegor-mialyk) [reply far/processname.cpp:194]: The line 258 has an extra check before accessing `str[si]`:
```cpp
if (si < str.size() && upper(pc) == upper(str[si]))
```

File: far/processname.cpp [hunk 2]
```diff
@@ -144,142 +144,132 @@ string ConvertWildcards(string_view const SrcName, string_view const Mask)
...
-	// BUGBUG rewrite
-
-	// Special case for these simplest and most common masks:
+	// Special case for these simplest and most common masks
 	if (pattern == L"*"sv || (CmpNameLegacyMode && pattern == L"*.*"sv))
 		return true;
 
 	if (pattern.empty() || str.empty())
-		return false;
+		return pattern.empty() && str.empty(); // used to be `false` but this looks more accurate
```
[the remark this answers is not in the document]
Comment (yegor-mialyk) [reply far/processname.cpp:154]: Yes, it is a change, easy to revert if needed.

File: far/processname.cpp [hunk 3]
```diff
@@ -144,142 +144,132 @@ string ConvertWildcards(string_view const SrcName, string_view const Mask)
...
-			/* $ 01.05.2001 DJ
-				оптимизированная ветка работает и для имен с несколькими
-				точками
-			*/
-			if (pattern[0] == L'.')
-			{
-				if (pattern.size() == 2 && pattern[1]==L'*')
-					return true;
+		return false;
+	};
```
[the remark this answers is not in the document]
Comment (yegor-mialyk) [reply far/processname.cpp:192]: All the checks are done before calling `match_char_in_set`, but it is possible to add an extra check here.

File: far/processname.cpp [hunk 4]
```diff
@@ -144,142 +144,132 @@ string ConvertWildcards(string_view const SrcName, string_view const Mask)
...
+			return false;
 
-		const auto stringc = str.empty()? 0 : upper(str.front());
-		const auto patternc = upper(pattern.front());
-		pattern.remove_prefix(1);
+		++StarSI;
+		si = StarSI;
+		pi = StarPI;
+		return true;
+	};
```
[the remark this answers is not in the document]
Comment (yegor-mialyk) [reply far/processname.cpp:172]: Same as before, `str[si]` is protected before accessing.

**Comment by alabuzhev** [comment]
> build 5274

Unfortunately I don't remember what it was about.
There's [this](https://forum.farmanager.com/viewtopic.php?t=11320) forum post close enough to that date, but renaming 0001.jpg to V*.* still gives V001.jpg, so perhaps it's unrelated.

> in some edge cases or with crazy patterns, it may produce slightly different results

Do you have any examples of such cases?

> mimic cmd is not a good idea

I'd say we should mimic whatever is common / logical / useful, not all the possible quirks of any command processor in particular.

**Comment by yegor-mialyk** [comment]
I did not touch the rename logic, so it works as before: renaming `0001.jpg` to `V*.*` gives `V001.jpg`.

> Do you have any examples of such cases?

Let's say you have the following  files:
```
test
test.
test..
```

If you search for `t*???.` you will get two files:
```
test.
test..
```

But if you search for `t???*.` you will get all three:
```
test
test.
test..
```

Technically both patterns (as well as something like this `t*?*?*?.`) are absolutely identical and all of them should potentially return the same result.

Both PowerShell and the new algorithm return two files for both patterns:
```
test.
test..
```

**Comment by alabuzhev** [comment]
> Hope the code is easy to read and understand.

It's easy to read indeed, yet still somewhat cryptic. But I suppose that's just the nature of the problem.
Let's merge it, I can't see why not.

A few minor comments:
- for the sake of consistency please drop trailing return types from lambdas where deducible and Egyptian braces
- add const where possible
- places like `upper(x) == upper(y)` can probably be replaced with https://github.com/FarGroup/FarManager/blob/fc29b7bc20fb35124ae5196f1711ac414b0d7e0f/far/string_utils.cpp#L188-L191
  - name matching is potentially a hot path and the `Chr1 == Chr2` branch might be slightly faster in common cases than going to kernel32 (e.g. typical masks and typical extensions are usually both in lower case anyway)
- build, changelog etc.

**Comment by alabuzhev** [comment]
Thanks

**Comment by alabuzhev** [comment]
The `ConvertWildcards` function in the same file should probably match `CmpName` logic, whatever that is.
I have attempted to assert that, but [without much success](https://github.com/FarGroup/FarManager/blob/d14fe119b109a9c7aebec688833c1846885093ab/far/processname.cpp#L465-L466).
Could you have a look please?
Should `CmpName` be more relaxed?
Or is `ConvertWildcards` just broken? (it probably is, since it [leaks the wildcards](https://github.com/FarGroup/FarManager/blob/d14fe119b109a9c7aebec688833c1846885093ab/far/processname.cpp#L452-L454) into the result in some cases)?
Or am I missing something obvious and that assertion doesn't make sense in the first place?

**Comment by yegor-mialyk** [comment]
I'll take a look.
